### ISSUE TITLE
[LLVM] 2 threads in java that execute fprintf in C in parallel cause illegal null pointer access exception

### ISSUE THREAD

**Issue by roxanan-adyen** [oracle/graal#4335]
**Describe GraalVM and your environment :**
 - GraalVM version: **22.0.0**
 - **EE**
 - JDK version: **JDK11**
 - OS and OS Version: **macOS bigSur**
 - Architecture: **amd64**
 - The output of `java -Xinternalversion`:
```
 **Java HotSpot(TM) 64-Bit Server VM (11.0.14+8-LTS-jvmci-22.0-b03) for bsd-amd64 JRE (11.0.14+8-LTS-jvmci-22.0-b03), built on Jan  2 2022 15:02:19 by "graal1" with gcc 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.17)**
```

**Have you verified this issue still happens when using the latest snapshot?**
Only tried it with 22.0.0

**Describe the issue**
I have a small C program where I call fprintf with 2 params in a loop. This is built as a library that is loaded by a java app. In the java app, I create two threads that will execute the C function. At some point, one of them get this exception:
```
Illegal null pointer access in 'store i8'.
```

**How to compile the LLVM bitcode that causes the problem**
 Compile C lib:
 ```
 $GRAALVM_MANAGED_LLVM_TOOLCHAIN/clang -fPIC -shared -O1 -g fprintf.c -lgraalvm-llvm -o crash.so
 ```
Where 
```
export GRAALVM_MANAGED_LLVM_TOOLCHAIN=$($GRAALVM_HOME/bin/lli --print-toolchain-path --llvm.managed)
```

**Code snippet or code repository that reproduces the problem**
C code: fprintf.c:
```
#include <stdio.h>

int my_function() {
        while (1) {
                fprintf(stdout, "%d %s\n", 1, "TestString");
                fflush(stdout);
        }

        return 0;
}
```
Java:
```
import java.io.*;
import org.graalvm.polyglot.*;

class Java2Threads {
        public static void main(String[] args) throws Exception {
                Context polyglot = Context.newBuilder().allowAllAccess(true).allowIO(true).option("llvm.managed", "true").build();
                File file = new File("crash.so");
                Source source = Source.newBuilder("llvm", file).build();
                Value cpart = polyglot.eval(source);

                Thread myThread = new Thread(() -> { // Lambda Expression
                        try {
                                System.out.println("MyThread running" + Thread.currentThread().getId());
                                cpart.invokeMember("my_function");
                        } catch (Exception e) {
                                System.out.println("Exception " + Thread.currentThread().getId());
                                e.printStackTrace();
                                System.exit(-1);
                        }
                });

                Thread myThread2 = new Thread(() -> { // Lambda Expression
                        try {
                                System.out.println("MyThread running" + Thread.currentThread().getId());
                                cpart.invokeMember("my_function");
                        } catch (Exception e) {
                                System.out.println("Exception " + Thread.currentThread().getId());
                                e.printStackTrace();
                                System.exit(-1);
                        }
                });

                myThread.start();
                myThread2.start();
                myThread.join();
                myThread2.join();
        }
}
```

**Steps to reproduce the problem**
Please include both build steps as well as run steps
1. $GRAALVM_MANAGED_LLVM_TOOLCHAIN/clang -fPIC -shared -O1 -g fprintf.c -lgraalvm-llvm -o crash.so
2. javac Java2Threads.java && java Java2Threads
 Output:
 ```
 ....
 1 TestString
Exception 19
Illegal null pointer access in 'store i8'.
        at <llvm> my_function(fprintf.c:5:55)
        at org.graalvm.sdk/org.graalvm.polyglot.Value.invokeMember(Value.java:932)
        at Java2Threads.lambda$main$0(Java2Threads.java:14)
        at java.base/java.lang.Thread.run(Thread.java:834)
1 TestString
```
        
**Expected behavior**
No exception, only the strings printed out (order doesn't matter).

**Additional context**
I also tried C++ version, with iostream and I get the same behavior.

**Comment by eregon** [comment]
It might be advisable to link your C library againt pthread (`-pthread`), even if it doesn't use it, otherwise C code/the compiler sometimes assume there will never be a second thread and then does some thread-unsafe things.

**Comment by roxanan-adyen** [comment]
@eregon I tried and it didn't help. I also put a comment in slack, but it may be helpful to put it here as well.

The story behind is that I have a more complex C/C++ library that will expose a function to the java side. In java, I would like to load the library once and then execute the function from any thread, any time. I don't have any global variable, so it should be fine. One issue I noticed is that java spawns user level threads that can map to the same pthread. Is this why fprintf doesn't work properly? And can this affect any other syscall from the C/C++ side?

**Comment by rschatz** [comment]
> One issue I noticed is that java spawns user level threads that can map to the same pthread.

In this scenario, there are two layers of pthreads: The host-level pthreads library that is used by GraalVM / the JVM to implement Java-level threads, and the guest-level pthreads library that is running inside GraalVM on top of the LLVM runtime.

The host-level pthreads should be completely irrelevant for us. That is an implementation detail of the VM. They only come into the picture in unmanaged mode, since you are running on managed mode, this should be fully abstracted and make no difference at all. As long as Java-level thread locals do what they are supposed to do, we don't care what host-level pthread our code is running on.

For the guest-level pthreads, there should be a 1:1 mapping between pthread and Java thread (at least for all the Java threads that are running LLVM code). If this is not the case, that's a bug on our side. Whether it causes that particular bug is another question.

I also wouldn't rule out a general thread-safety issue in our runtime.

I'll investigate.

**Comment by rschatz** [comment]
I can confirm that both threads in this example really return the same value for `pthread_self()`. This is definitely a bug.

Not 100% sure, but this is likely the reason for this null pointer access. In the libc code, concurrent access to file descriptors is synchronized by CAS-ing the thread id into the "lock" field of the file descriptor. If two threads have the same id, that won't work obviously.

I still haven't found the reason why. It seems to happen really reliably in this example. The weird thing is that this test in our repo does *not* fail, and it looks pretty much the same as your example: https://github.com/oracle/graal/blob/d1fccad4a6f71c4ca93b6112e86e8d48f69521c1/sulong/tests/com.oracle.truffle.llvm.tests/src/com/oracle/truffle/llvm/tests/interop/ThreadsInteropTest.java

**Comment by roxanan-adyen** [comment]
Thanks a lot for looking into this @rschatz! Not sure it it matters, but in your test, I don't see the option "llvm.managed" set to true.

**Comment by rschatz** [comment]
This should be set by the test runner. The same test is used for both managed and unmanaged mode. I hope at least. There could be a bug in the test runner of course.

**Comment by rubyFeedback** [comment]
It's like an onion, potentially! Bugs concealing other layers of bugs on other bugs! \o/

**Comment by rschatz** [comment]
> I get the same pthread address. But what I noticed, if I use gettid(), I get indeed a different id per thread

Huh, that's really weird, because `gettid()` is implemented like this:
```
pid_t gettid(void)
{
	return __pthread_self()->tid;
}
```

**Comment by rschatz** [comment]
I made some progress debugging, and I think I found at least the `printf` synchronization issue.

Apparently, musl libc contains an optimization: It's not doing any synchronization on file datastructures at all if `pthread_create` was never called, assuming the whole application is single-threaded.

There is a little block of code in `pthread_create` that lazily initializes all locks if this wasn't done before. We have to replicate the effect of that code also if a managed thread is created, to keep the state consistent.

I did test that with both my unittest and with your example, and it seems to work.

As a workaround, can you try just calling `pthread_create` once at application startup? You can just create a dummy thread that immediately terminates, shouldn't matter, as long as its called once before creating any other threads, it should fix the problem. That's of course only a workaround and no substitute for fixing the bug on our side, but I think it's the quickest way for you to verify whether this is really the issue.

**Comment by roxanan-adyen** [comment]
I did try calling `pthread_create` and indeed, I don't see the crash anymore. 

And this is very confusing. 
```
pid_t gettid(void)
{
	return __pthread_self()->tid;
}
``` 

This is what I get when I run 2 threads. Printing is done only by java. I have two function, one returns pthread_self and returns gettid value. This is the output:
```
Java tid 20
Java tid 19
Gettid 19
Gettid 20
Pthread_self ManagedMemoryBlock+0x18
Pthread_self ManagedMemoryBlock+0x18
```

**Comment by rschatz** [comment]
`ManagedMemoryBlock+0x18`

That's just the default `toString` of a pointer to a managed memory block, pointing to an offset of `0x18` bytes into some block of managed memory. It doesn't tell you whether it's the same memory block. But `Value.equals` should return false for these two pointers.

I guess we should at least print out some sort of object id or identity hash. It's not going to be collision free, but I guess for debug output it should be enough.

**Comment by roxanan-adyen** [comment]
@rschatz My mistake. Indeed, they are different. When I first assumed both threads are pointing to the same structure I used printf from C side, and that was misleading. I think they are indeed different.

**Comment by roxanan-adyen** [comment]
@rschatz I created another ticket for the deadlock issue. https://github.com/oracle/graal/issues/4349

**Comment by rschatz** [comment]
This is fixed on the master branch, the fix will be in the next release (22.1). I did also fix the `toString` output for managed pointers, it will now produce output that makes it possible to see whether it's the same memory block or a different one.
